### PR TITLE
[WIP][Build] Add repl mode to launch LLDB REPL

### PR DESCRIPTION
**Pull request by bhargavg**
This patch adds `repl` sub command to `swift build` that can launch REPL with all the modules linked.

Dependencies:

- [x] Merging of #388

### CODE REVIEW

File: Sources/Commands/SwiftBuildTool.swift [hunk 1]
```diff
@@ -157,6 +162,28 @@ public struct SwiftBuildTool {
...
+                let (modules, externalModules, _) = try transmute(rootPackage, externalPackages: externalPackages)
+
+                /// TODO: Need to create DSOs for C executables as well.
+                let products: [Product] = modules.flatMap{ $0 as? SwiftModule }
+                                                 .map{ return Product(name: $0.name, type: .Library(.Dynamic), modules: [$0]) }
+
+                let yaml = try describe(opts, .debug, modules, Set(externalModules), products, toolchain: toolchain)
+                try build(YAMLPath: yaml, target: opts.buildTests ? "test" : nil)
+
+                var args = ["-L", ".build/debug", "-I", ".build/debug"]
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftBuildTool.swift]: I think this should be `\(opts.buildPath)/debug`

File: Sources/Commands/UserToolchain.swift [hunk 2]
```diff
@@ -34,10 +35,16 @@ struct UserToolchain: Toolchain {
...
     init() throws {
         do {
+
             SWIFT_EXEC = getenv("SWIFT_EXEC")
                 // use the swiftc installed alongside ourselves
                 ?? Path.join(Process.arguments[0], "../swiftc").abspath
 
+            lldb = getenv("LLDB")
+                // use the lldb installed alongside ourselves
+                ?? Path.join(Process.arguments[0], "../lldb").abspath
```
Comment (bhargavg) [inline Sources/Commands/UserToolchain.swift]: @ddunbar Can we expect this env var to be set?

AFAIK `Path.join(Process.arguments[0], "../lldb").abspath` will fail, because we are not creating a symlink to `lldb`.

File: Sources/Commands/SwiftBuildTool.swift [hunk 3]
```diff
@@ -157,6 +162,28 @@ public struct SwiftBuildTool {
             }
         
             switch mode {
+            case .repl(let toolchain):
+                let (rootPackage, externalPackages) = try fetch(opts.path.root)
+                let (modules, externalModules, _) = try transmute(rootPackage, externalPackages: externalPackages)
+
+                /// TODO: Need to create DSOs for C executables as well.
+                let products: [Product] = modules.flatMap{ $0 as? SwiftModule }
+                                                 .map{ return Product(name: $0.name, type: .Library(.Dynamic), modules: [$0]) }
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftBuildTool.swift]: nit: return can be omitted

File: Sources/Commands/UserToolchain.swift (hunk 2 above)
Comment (aciidb0mb3r) [inline Sources/Commands/UserToolchain.swift]: why not just use `swift`?

Comment (bhargavg) [inline Sources/Commands/UserToolchain.swift]: We only have path for `swiftc` and not `swift` in `Toolchain`. So a path needs to be added either for invoking REPL.

I looked at how swift is launching repl, it's nothing but `lldb --repl`. So I added path for `lldb` directly.

Comment (aciidb0mb3r) [inline Sources/Commands/UserToolchain.swift]: ah

Comment (ddunbar) [inline Sources/Commands/UserToolchain.swift]: I would rather call `swift` to launch the REPL than depend on how lldb is launched.

File: Sources/Commands/SwiftBuildTool.swift [hunk 4]
```diff
@@ -157,6 +162,29 @@ public struct SwiftBuildTool {
...
+                var args = ["-L", libdir, "-I", libdir]
+                args += toolchain.platformArgs
+                args += modules.map{ return "-l\($0.name)" }
+
+                let finalCommand = [toolchain.lldb, "--repl=" + args.joined(separator: " ") ]
+
+                /// Get the $TERM environment variable.
+                /// This is needed by lldb to start the REPL.
+                let term = getenv("TERM") ?? "dumb"
+                try system(finalCommand, environment: ["TERM": term])
```
Comment (ddunbar) [inline Sources/Commands/SwiftBuildTool.swift]: It probably makes sense to launch `swift` inheriting the full environment in this mode, not specific things we handpick.

File: Sources/Commands/SwiftBuildTool.swift [hunk 5]
```diff
@@ -157,6 +162,29 @@ public struct SwiftBuildTool {
...
             switch mode {
+            case .repl(let toolchain):
+                let (rootPackage, externalPackages) = try fetch(opts.path.root)
+                let (modules, externalModules, _) = try transmute(rootPackage, externalPackages: externalPackages)
+
+                /// TODO: Need to create DSOs for C executables as well.
+                let products: [Product] = modules.flatMap{ $0 as? SwiftModule }
+                                                 .map{ Product(name: $0.name, type: .Library(.Dynamic), modules: [$0]) }
+
+                let yaml = try describe(opts, .debug, modules, Set(externalModules), products, toolchain: toolchain)
```
Comment (ddunbar) [inline Sources/Commands/SwiftBuildTool.swift]: If this mode requires us to build additional stuff, I would prefer that be modeled as an extra argument to the describe() function selecting the mode, and then have that function embed the implementation details of what extra things need to be done to support that mode (i.e., the computation of other products)

File: Sources/PackageModel/Product.swift [hunk 6]
```diff
@@ -41,11 +41,6 @@ public class Product {
     }
 }
 
-public enum LibraryType {
-    case dynamic
-    case Static
-}
-
```
Comment (ddunbar) [inline Sources/PackageModel/Product.swift]: You are free to land these two diffs on master directly if they are really just trivial refactors (or land in a separate PR you send through test-and-merge).

File: Sources/Commands/SwiftBuildTool.swift [hunk 7]
```diff
@@ -25,18 +25,22 @@ import enum Utility.ColorWrap
...
 private enum Mode: Argument, Equatable, CustomStringConvertible {
     case build(Configuration, Toolchain)
     case clean(CleanMode)
+    case repl(Toolchain)
     case usage
     case version
 
     init?(argument: String, pop: () -> String?) throws {
         switch argument {
+        case "repl":
```
Comment (ddunbar) [inline Sources/Commands/SwiftBuildTool.swift]: I think I would prefer to call this `--repl` for now, until we know exactly where we want it to live.

**Comment by ddunbar** [comment]
Sweet, thanks for working on this! This will be an awesome feature...

If I try this commit on SwiftPM itself, I get a failure:
```
$ swift-dev build repl
Compile Swift Module 'PackageDescription' (6 sources)
Compile Swift Module 'swiftpm_xctest_helper' (1 sources)
Compile Swift Module 'libc' (2 sources)
Linking .build/debug/liblibc.dylib
Compile Swift Module 'POSIX' (15 sources)
Linking .build/debug/libPackageDescription.dylib
Compile Swift Module 'PackageDescriptionTestSuite' (3 sources)
Linking .build/debug/libswiftpm-xctest-helper.dylib
<unknown>:0: error: module name "swiftpm-xctest-helper" is not a valid identifier; use -module-name flag to specify an alternate name
<unknown>:0: error: build had 1 command failures
error: exit(1): /Volumes/Data/ddunbar/public/swift-project/build/Ninja-ReleaseAssert/swift-dev.xctoolchain/usr/bin/swift-build-tool -f /Volumes/Data/ddunbar/public/swift-project/swiftpm/.build/debug.yaml
```

Can you investigate?

File: Sources/PackageModel/Product.swift [hunk 8]
```diff
@@ -14,9 +14,9 @@ import Utility
 public class Product {
     public let name: String
     public let type: ProductType
-    public let modules: [SwiftModule]
+    public let modules: [Module]
 
-    public init(name: String, type: ProductType, modules: [SwiftModule]) {
```
Comment (bhargavg) [inline Sources/PackageModel/Product.swift]: This is already handled in #388. Waiting for it to get merged first.

**Comment by bhargavg** [comment]
@ddunbar I've fixed the issue that caused repl to fail when invoked on SwiftPM. Now I'm facing another problem.

```bash
$ swift -L .build/debug/ -I .build/debug -llibc -lPOSIX -lUtility -lBasic -lSourceControl
Welcome to Apple Swift version 3.0-dev (LLVM 8fcf602916, Clang cf0a734990, Swift 000d413a62). Type :help for assistance.
  1> import SourceControl
warning: (x86_64) /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/liblibc.dylib empty dSYM file detected, dSYM was created with an executable with no debug info.
objc[26502]: Class _TtC5Basic16OutputByteStream is implemented in both /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libBasic.dylib and /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libSourceControl.dylib. One of the two will be used. Which one is undefined.
objc[26502]: Class _TtC5Basic13TOMLItemArray is implemented in both /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libBasic.dylib and /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libSourceControl.dylib. One of the two will be used. Which one is undefined.
objc[26502]: Class _TtC5Basic13TOMLItemTable is implemented in both /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libBasic.dylib and /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libSourceControl.dylib. One of the two will be used. Which one is undefined.
objc[26502]: Class _TtC7Utility3Git is implemented in both /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libUtility.dylib and /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libSourceControl.dylib. One of the two will be used. Which one is undefined.
objc[26502]: Class _TtC7Utility25StandardErrorOutputStream is implemented in both /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libUtility.dylib and /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libSourceControl.dylib. One of the two will be used. Which one is undefined.
objc[26502]: Class _TtC7UtilityP33_85EDAD19EFE38364111E13D20EB47E9526DirectoryContentsGenerator is implemented in both /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libUtility.dylib and /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libSourceControl.dylib. One of the two will be used. Which one is undefined.
objc[26502]: Class _TtC7Utility36RecursibleDirectoryContentsGenerator is implemented in both /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libUtility.dylib and /Users/bhargavg/Documents/workspaces/xcode/github/apple/swiftpm/.build/debug/libSourceControl.dylib. One of the two will be used. Which one is undefined.
  2>
```

This is happening because `libUtility.dylib` is being statically linked when `libSourceControl.dylib` is built causing duplicate symbols:
```bash
$ nm -gU .build/debug/libSourceControl.dylib | grep "Utility.Git$"
000000000009dd70 D __TMC7Utility3Git
000000000009e858 S __TMLC7Utility3Git
0000000000044740 T __TMaC7Utility3Git
000000000009dd38 D __TMmC7Utility3Git
0000000000092cc0 S __TMnC7Utility3Git
$ nm -gU .build/debug/libUtility.dylib | grep "Utility.Git$"
000000000005a6f0 D __TMC7Utility3Git
000000000005af98 S __TMLC7Utility3Git
00000000000088c0 T __TMaC7Utility3Git
000000000005a6b8 D __TMmC7Utility3Git
00000000000551b0 S __TMnC7Utility3Git
```

How can we avoid this?

I thought about making a single product with all the `.dylib` targets, but this will fail incase of statically linked executables.

**Comment by ddunbar** [comment]
For this to work properly, we probably need to be creating a single dylib per target and linking them together amongst themselves properly. This is the compilation model that is currently primarily used with Swift. We have been playing with static linking which has a number of benefits, but not following the standard model causes problems.

This is very much orthogonal to this PR, though, so I think we should either fix that first, or if they aren't going to conflict then land this and just not have it work completely until that other work is done.

I have one other request on this PR which I will add inline...

File: Sources/Commands/SwiftBuildTool.swift (hunk 5 above)
Comment (ddunbar) [inline Sources/Commands/SwiftBuildTool.swift]: I thought about this more, and what I would like to do instead here is ensure that the default `build.yaml` has targets for all of the situations we might want to build for (similar to what we do for tests), but not all of them need to be built by default. So we shouldn't need to regenerate the build file just for this option.

Does that make sense?

**Comment by bhargavg** [comment]
@ddunbar Strange, I couldn't replicate the "duplicate symbols" issue outside SwiftPM.

Could you please check this when you have time?
[TestSPMREPL.zip](https://github.com/apple/swift-package-manager/files/304590/TestSPMREPL.zip)

To invoke REPL:

```bash
$ export LLDB=path/to/lldb
$ .build/debug/swift-build --repl
```

**Comment by ddunbar** [comment]
Misclick, didn't mean to close.

**Comment by bhargavg** [comment]
@ddunbar As per your comments above, I've made the following changes:

1. Add `repl` target in `build.yaml` with all the possible products.
2. Add `--repl` flag to launch REPL. This will invoke `swift-build-tool` with `repl` as target.
3. Launch REPL directly through `swift` instead of `lldb`.

File: Sources/Build/describe().swift [hunk 9]
```diff
@@ -60,9 +71,14 @@ public func describe(_ prefix: String, _ conf: Configuration, _ modules: [Module
         }
     }
 
-    for product in products {
+    /// TODO: handle system modules here.
+    let productsForREPL = modules.filter{ !($0 is CModule) }.map{ Product(name: $0.c99name, type: .Library(.Dynamic), modules: [$0]) }
```
Comment (bhargavg) [inline Sources/Build/describe().swift]: @ddunbar Thoughts on how to handle system modules?

**Comment by aciidb0mb3r** [comment]
Closing due to inactivity, however feel free to open a new one!
